icu4c/source/common/localematcher.cpp

```diff
@@ -446,21 +447,32 @@ LocaleMatcher::LocaleMatcher(const Builder &builder, UErrorCode &errorCode) :
                         supportedLSRs[paradigmIndex] = &lsr;
                         supportedIndexes[paradigmIndex++] = i;
                     } else {
-                        supportedLSRs[otherIndex] = &lsr;
-                        supportedIndexes[otherIndex++] = i;
+                        supportedLSRs[--otherIndex] = &lsr;
+                        supportedIndexes[otherIndex] = i;
                     }
                 }
             }
             if (U_FAILURE(errorCode)) { return; }
         }
-        // Squeeze out unused array slots.
-        if (paradigmIndex < paradigmLimit && paradigmLimit < otherIndex) {
-            uprv_memmove(supportedLSRs + paradigmIndex, supportedLSRs + paradigmLimit,
-                         (otherIndex - paradigmLimit) * sizeof(const LSR *));
-            uprv_memmove(supportedIndexes + paradigmIndex, supportedIndexes + paradigmLimit,
-                         (otherIndex - paradigmLimit) * sizeof(int32_t));
+        // Reverse the non-paradigm LSRs to be in order, right after the paradigm LSRs.
+        // First fill the unused slots between paradigm LSRs and other LSRs.
+        // This gap is as large as the number of locales with duplicate LSRs.
+        int32_t i = paradigmIndex;
+        int32_t j = supportedLocalesLength - 1;
+        while (i < otherIndex && otherIndex <= j) {
+            supportedLSRs[i] = supportedLSRs[j];
+            supportedIndexes[i++] = supportedIndexes[j--];
         }
-        supportedLSRsLength = otherIndex - (paradigmLimit - paradigmIndex);
+        // Swap remaining non-paradigm LSRs in place.
```

Comment:
If you always had a gap of at least one (which you could set up at the beginning), you wouldn't need the temps, which would save a (very slight) amount of time).

Reply:
True, but it would leave a gap in the middle. I would then have to either move the trailing section up to close the gap, or null the gap and modify the LocaleDistance code to skip over it (and toString() to not print it). Not worth it I think.